[do not merge] supervisor for kafka indexing tasks

dclim opened the pull request:
This depends on PR #2220.

navis committed:
Support cardinality for search query (Fix for #2260)

himanshug committed:
at broker - only add segments from specific tiers to the timeline

dclim committed:
supervisor for kafka indexing tasks

dclim commented:
Short-term TODOs for MVP+1:
- when a task succeeds, instead of killing the other running replicas forcefully and generating failures, issue an RPC to the peon to instruct it to stop early with a success status
- if we have tasks that keep failing consistently, should we just stop the supervisor and report an error?
- query the peons directly to find out their current offset and start the next task early to avoid ingestion pauses during segment building and handoff
- store the supervisor job in metadata storage (?) and load these jobs on supervisor startup to support overlord failovers and restarts without needing to resubmit jobs

pjain1 committed:
remove extra kill parameter

pjain1 committed:
better exception for invalid interval

himanshug committed:
update IncrementalIndex to support unsorted facts map that can be used in groupBy merging to improve performance

himanshug committed:
while GroupBy merging use unsorted facts in IncrementalIndex wherever possible

drcrallen committed:
Make CloserRule use guava's Closer

gianm commented on extensions/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,137 @@
+/**
+ * Manages the creation and execution of {@link KafkaSupervisorWorker}.
+ */
+public class KafkaSupervisor implements IngestionSupervisor<KafkaSupervisorJob>
```
This doesn't seem to be doing really Kafka-specifc stuff… it seems like it'd make more sense to:
- Move this class to core and call it "SupervisorManager"
- Rename "Worker" to "Supervisor" (worker already means something else in the indexing service anyway)

The advantage of that being that modules that want to implement supervision don't need to have a class like this one, as the SupervisorManager in core would handle management of all of them.

gianm replied:
Another advantage is that in the future, we'll want to remember which supervisors should be running and to start them back up again, and keeping that state seems like a good job for a core SupervisorManager.

dclim replied:
Sounds good, in that case I can also move the Jersey resource into core which would simplify things.

dclim replied:
@gianm refactored KafkaSupervisor to be a generic SupervisorManager in core and renamed a bunch of stuff

gianm committed:
Plumbers: Move plumber.add out of try/catch for ParseException.

The incremental indexes handle that now so it's not necessary.

Also, add debug logging and more detailed exceptions to the incremental
indexes for the case where there are parse exceptions during aggregation.

gianm committed:
examples: Switch druid-server, druid-common to "provided".

gianm committed:
ThreadPoolTaskRunner: Make graceful shutdown logs less scary.

- It's okay to suppress InterruptedException during graceful shutdown, as
  tasks may use it to accelerate their own shutdown.
- It's okay to ignore return statuses during graceful shutdown (which may
  be FAILED!) because it actually doesn't matter what they are.

fjy committed:
Merge pull request #2632 from gianm/examples-druid-provided

examples: Switch druid-server, druid-common to "provided".

gianm commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorConfig.java:
```diff
@@ -0,0 +1,35 @@
+
+public abstract class SupervisorConfig
+{
+  @Config("druid.indexer.supervisor.numThreads")
```
Usually in druid we use JsonConfigProvider rather that a Skife based config.

gianm commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -0,0 +1,134 @@
+  public String createSupervisor(SupervisorSpec spec)
+  {
+    if (!started) {
+      throw new IllegalStateException("Supervisor must be started before it can accept jobs.");
```
SupervisorManager

gianm committed:
WorkerTaskMonitor: stop() waits for mainLoop to exit.

Fixes #2637.

Commit:
fix broken link for Tasks

himanshug committed:
Merge pull request #2640 from guobingkun/fix_doc

fix broken link for Tasks

fjy committed:
Merge pull request #2638 from gianm/attempt-fix-2637

WorkerTaskMonitor: stop() waits for mainLoop to exit.

xvrl committed:
optimize timeboundary for min or max bound

drcrallen committed:
Merge pull request #2631 from gianm/plumbers-rpe

Better logging for ParseExceptions on index aggregation, and remove unnecessary exception handling.

nishantmonu51 committed:
Merge pull request #2634 from gianm/stopGracefully-avoid-interrupt

ThreadPoolTaskRunner: Make graceful shutdown logs less scary.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+  private static final EmittingLogger log = new EmittingLogger(KafkaSupervisor.class);
+  private static final Random RANDOM = new Random();
+
+  protected final String supervisorRunId;
```
This appears to be protected so tests can access it. package-private is good enough for that; "protected" makes it sound like it's for subclasses (protected is the same as package-private except it allows subclasses too).

Also if this is going to be read only (that's what it looks like) then it'd be better to have it be private, then have a package-private getter with `@VisibleForTesting` on it.

dclim replied:
Thanks, I meant to make it package-private!

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+    boolean initialized = false;
+  }
+
+  protected static final String SUPERVISOR_CONTEXT_KEY = "supervisorId";
```
It looks like this is just used in the tests, it could be package-private and still work there.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+          ImmutableMap.<String, Object>of(SUPERVISOR_CONTEXT_KEY, supervisorRunId)
+      );
+
+      if (taskMaster.getTaskQueue().isPresent()) {
```
It's better to grab the Optional, check it, then `get()` it. Otherwise you're vulnerable to a race where you lose leadership between the call to `isPresent()` and the call to `get()`.

gianm replied:
Similar comment applies to other uses of TaskMaster Optionals here.

dclim replied:
Good tip, thanks

gianm commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorResource.java:
```diff
@@ -0,0 +1,161 @@
+          @Override
+          public Response apply(SupervisorManager manager)
+          {
+            if (!manager.getSupervisor(id).isPresent() || !manager.getSupervisor(id)
```
Grab the Optional first, then check & use it.

gianm commented on indexing-service/src/main/java/io/druid/indexing/overlord/supervisor/SupervisorManager.java:
```diff
@@ -0,0 +1,85 @@
+    return supervisors;
+  }
+
+  public Optional<Supervisor> getSupervisor(String id)
```
It looks like all the callers of this are actually just wanting the spec, so it should simplify things to do:
- Change this to `getSupervisorSpec`
- Remove `getSpec` from Supervisor (I don't think anything would need that anymore)

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+             .emit();
+        }
+
+        Thread.sleep(config.getRunPeriod());
```
minor: I think usually "run every X" is easier to reason about than "sleep for X every run". You could get that by remembering the start time of each loop and then doing some math on that.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+      return;
+    }
+
+    List<PartitionInfo> partitions = consumer.partitionsFor(spec.getIoConfig().getTopic());
```
`listTopics` returns the partitionInfos for each topic as a Map, right? Seems like it'd be simpler to pull the partitionInfos out of that map. Or, just skip the call to `listTopics` and use `partitionsFor` instead.

Either way, seems like we don't need both…

dclim replied:
I was finding that partitionsFor() didn't actually detect new partitions that were made, even when I had metadata.max.age.ms set, it was still returning cached data. I just confirmed it now by looking at the implementation. Calling listTopics() results in a call to the server to update the metadata. You're right, the partitionsFor() call is unnecessary.

gianm replied:
ah, okay, it looks like the background refresh only happens on `poll()`. Okay cool.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+
+  private volatile boolean started = false;
+
+  private final HashMap<Integer, TaskGroup> taskGroups = new HashMap<>();
```
Comment about what the key means here? I think it's a task group id.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+
+  private void killUnownedKafkaTasks()
+  {
+    if (taskMaster.getTaskQueue().isPresent()) {
```
Grab optional first, then check.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+      for (Task task : tasks) {
+        if (task instanceof KafkaIndexTask
+            && dataSource.equals(task.getDataSource())
+            && !supervisorRunId.equals(task.getContextValue(SUPERVISOR_CONTEXT_KEY))) {
```
Hmm, this looks like it's going to kill any task that wasn't started by us. That's more aggressive than we really want, since it means tasks will be dying during a normal restart of overlords. Ideally the tasks shouldn't die.

So it'd be better to let the tasks live if they're doing something we approve of. I guess we can talk about what exactly that means but I think a place to start thinking about that would be:
- The replication stuff works properly if the tasks create the same segments. So that at least means we want the same dataSchema, startPartitions, endPartitions, and maxRowsPerSegment. Possibly other stuff too.
- If our instances of those things are the same as what's already running, we shouldn't need to kill what's running.
- If our instances of those things are _not_ the same as what's already running, we could kill what's running, or we could let it stay and just adjust how we replace things (i.e. replace dead tasks in existing task groups with tasks that are "compatible" with existing ones, rather than using our templates for new tasks)

This may be something we want to address now or may be something we want to file a follow-on issue for…

dclim replied:
Hmm yeah, the 'killing any task that wasn't started by us' behavior came out of discussions of the possibility of two overlords both thinking they are the leader and spawning a bunch of Kafka indexing tasks that do the same thing. I think the concern was that there'd be a scenario where they didn't generate the same sequenceNames (somehow were using different starting offset ranges or a different ingestion spec) and then we'd get data duplication, so if I remember correctly we thought it might make sense for new supervisors to just kill the other overlord's tasks until the split-leader state settled. 

I like the idea of taking a more intelligent look at the running tasks and seeing if they're doing what we want and only taking action if they don't match our expectation.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+    }
+  }
+
+  private class TaskGroup
```
Some comments about what these fields mean, please. This class seems important!

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+        taskGroups.put(taskGroupId, new TaskGroup());
+      }
+
+      if (!taskGroups.get(taskGroupId).partitions.containsKey(partition)) {
```
this code would be a little simpler if you kept a reference to `taskGroups.get(taskGroupId)`

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+        for (Integer partition : kafkaTask.getIOConfig().getStartPartitions().getPartitionOffsetMap().keySet()) {
+          if (taskGroupId == null) {
+            taskGroupId = getTaskGroupIdForPartition(partition);
+          } else if (taskGroupId != getTaskGroupIdForPartition(partition)) {
```
Use `.equals` with boxed Integers, `==` and `!=` don't always work.

gianm replied:
Apparently it works for "small" ints.

``` scala
scala> Int.box(32) eq Int.box(32)
res0: Boolean = true

scala> Int.box(320) eq Int.box(320)
res1: Boolean = false
```

[eq in scala is like == in java]

Not something you want to rely on!

dclim replied:
Ah good catch, yeah it works fine for -128 to 127 because it caches those objects:

https://www.owasp.org/index.php/Java_gotchas#Immutable_Objects_.2F_Wrapper_Class_Caching

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+  private void updateTaskMetadata()
+  {
+    int taskCount = 0;
+    List<Task> tasks = taskStorage.getActiveTasks();
```
hmm, this may be easier to reason about if we grab a list of active tasks at the start of each `mainLoop` iteration, and then remove things and add things as we kill and create them? maybe?

dclim replied:
That's basically what it's doing - the only complication is that instead of operating on just the list of active tasks, it adds them to its list of "tracked tasks" and then removes and adds things from that list.. the reason being that when tasks succeed or fail they are no longer reported in activeTasks so we need to track their IDs independently so that we can query their status and know to either move on to the next set of offsets or retry a failed task.

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+          } else if (taskGroupId != getTaskGroupIdForPartition(partition)) {
+            log.error("Discovered task [%s] has different partition allocation than expected!", task.getId());
+            taskGroupId = null;
+            break;
```
What's the rationale behind `break`ing here? It looks like we'll move on to the next task (this breaks out of the `partitions` loop).

dclim replied:
The rationale is that if it finds a task that's handling a different combination of partitions than what it expects (which means the task was created with a spec that had a different 'taskCount' attribute or the partition assignment algorithm changed), it would skip adding that task to the map of tracked tasks and just ignore it. The thought behind it was that a different set of partitions means a different sequenceName which puts us at risk for ingesting duplicate data, but if that's true, maybe it actually would make more sense to kill that task rather than just logging an error. With the current implementation (aggressively killing all KafkaIndexTasks for our dataSource that were not created by us) I don't think we would ever actually encounter this case.

gianm replied:
ah okay, that makes sense. Yeah, I was wondering why a task would get flagged here but _not_ be killed earlier.

gianm replied:
If we're really expecting this case to never happen, perhaps instead of breaking out of this one task, we should instead emit an alert and throw an exception (aborting the entire management run)

gianm commented on extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/supervisor/KafkaSupervisor.java:
```diff
@@ -0,0 +1,617 @@
+
+  private void killRunningTasksInGroup(int taskGroupId)
+  {
+    if (taskMaster.getTaskQueue().isPresent()) {
```
Grab Optional, then check/get.

erikdubbelboer committed:
Update com.maxmind.geoip2 to 2.6.0

com.maxmind.geoip2 2.6.0 depends on com.google.http-client 1.15.0-rc (3 years old).
When trying to include other libraries in Druid that require an up to date version of com.google.http-client this causes a problem.

sderosiaux committed:
Ignore case when testing for table existence

b-slim committed:
Merge pull request #2644 from metamx/optimize-timeboundary

optimize timeboundary for min or max bound

rasahner committed:
Merge pull request #2540 from pjain1/remove_kill

Remove extra parameter from deleteDataSourceSpecificInterval endpoint and correct exception message for invalid interval

himanshug committed:
Merge pull request #2220 from gianm/appenderator-kafka

Appenderators, DataSource metadata, KafkaIndexTask

nishantmonu51 committed:
Merge pull request #2646 from atomx/update-maxmind

Update com.maxmind.geoip2 to 2.6.0

gianm committed:
Docs on default ports.

fjy committed:
Merge pull request #2260 from navis/cardinality-for-searchquery

Support cardinality for search query

drcrallen committed:
Merge pull request #2648 from chtefi/master

Ignore case when testing for table existence

fjy committed:
Merge pull request #2651 from gianm/ports-docs

Docs on default ports.

fjy committed:
Merge pull request #2571 from himanshug/gp_by_avoid_sort

avoid sort while doing groupBy merging when possible

jon-wei committed:
Merge pull request #2382 from himanshug/broker_segment_tier_selection

at broker, if configured, only add segments from specific tiers to the timeline

drcrallen committed:
Merge pull request #2579 from metamx/closerIsCloser

Make CloserRule use guava's Closer
